xc/xenops_server_xen.ml

```diff
@@ -448,24 +448,87 @@ end
 (* We store away the device name so we can lookup devices by name later *)
 let _device_id kind = Device_common.string_of_kind kind ^ "-id"
 
-(* Return the xenstore device with [kind] corresponding to [id] *)
+(* Return the xenstore device with [kind] corresponding to [id]
+
+This is an inefficient operation because xenstore indexes the devices by devid, not id,
+so in order to find an id we need to go through potentially all the devids in the tree.
+
+Therefore, we need to cache the results to decrease the overall xenstore read accesses.
+During VM lifecycle operations, this cache will reduce xenstored read accesses from
+O(n^2) to O(n), where n is the number of VBDs in a VM.
+*)
+
+module DeviceCache = struct
+	module PerVMCache = struct
+		include Hashtbl
+		let create n = (create n, Mutex.create ())
+	end
+	include Hashtbl
+	let create n = (create n, Mutex.create())
+	let discard (cache, mutex) domid = Mutex.execute mutex (fun () ->
```

Comment:
there are 2 levels of hashtables:
- the lowest level is a hashtable that keeps the cached devices for a specific domid
- the highest level is a hashtable the keeps the lowest level hashtables

The create for each of the levels has been redefined to return not only what create would return, but also a unique mutex for that hashtable instance.